update to use with auto-generated json

diivm opened the pull request.

diivm committed:
[generate_bindings.py] progress

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+
+class kind:
+
+    KIND = None
```
What's the difference between the `VARIABLES` and the `variables`?

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+    state_stack = []
+    kind_functions = {
+        "TRANSLATION_UNIT": [],
+        "NAMESPACE": [handle_NAMESPACE_0, handle_NAMESPACE_1],
```
Why not `handle_namespace_0`? In Python either use `ALL_CAPS_SNAKE_CASE` or `all_non_caps_snake_case`

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+
+        self.state_stack.append({"kind": KIND, "name": NAME, "depth": DEPTH})
+
+        if self.kind_functions[KIND][0]:
```
Possible error if the error has no members (also down below at index 1, L42)

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+            self.kind_functions[KIND][0](item)
+
+        for item in MEMBERS:
+            self.handle_kind(item)
```
You might want to rename `handle_kind` since it's handling child members instead of `kind`

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+
+    def handle_STRUCT_DECL_0(self, item):
+        MEMBERS = item["members"]
+        for item in MEMBERS:
```
Don't shadow variable names

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+            if item["kind"]=="CXX_BASE_SPECIFIER":
+                CXX_BASE_SPECIFIER = item["name"]
+                self.linelist.append(f'py::class_<{self.NAME, CXX_BASE_SPECIFIER}>(m, "{self.NAME}")')
+                return
```
You're handling just 1 base class (and only the base class in this function)? Please note so with a comment

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+
+    def handle_STRUCT_DECL_0(self, item):
+        MEMBERS = item["members"]
+        for item in MEMBERS:
+            if item["kind"]=="CXX_BASE_SPECIFIER":
+                CXX_BASE_SPECIFIER = item["name"]
+                self.linelist.append(f'py::class_<{self.NAME, CXX_BASE_SPECIFIER}>(m, "{self.NAME}")')
+                return
+        self.linelist.append(f'py::class_<{self.NAME}>(m, "{self.NAME}")')
```
I'd suggest restructuring it to be like:
```python
if has_base_speficier:
  do_this()
else:
  do_that()
```

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -1,4 +1,82 @@
+    NAME = None
+    MEMBERS = None
+    DEPTH = None
+    state_stack = []
```
Seems like this should be `_state_stack` instead since the user should not modify this

kunaltyagi commented on bindings/python/generate_bindings.py:
```diff
@@ -36,71 +110,86 @@ def handle_final(filename, module_name):
+
+def handle_PARM_DECL(item):
+    for sub_item in item["members"]:
+        kind_functions[sub_item["kind"]](sub_item)
```
Where is `kind_functions` at global scope?

diivm replied:
Changed to a class member

kunaltyagi reviewed (commented):
partial

divmadan commented:
@kunaltyagi 
Thanks for the review. I must point out that this code needs cleaning, and I haven't even run this code as of yet. Just typing my mind now. :slightly_smiling_face: 
It will indicate you shortly when it is ready for review, else it will waste your time.

divmadan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [generate pybind code for point_types by using json generated via libclang.py](https://trello.com/c/cNKU40aN/31-generate-pybind-code-for-pointtypes-by-using-json-generated-via-libclangpy)

diivm committed:
some more progress

diivm committed:
still working

diivm committed:
some more

diivm committed:
a little more

diivm committed:
some more

diivm committed:
it generates!!

diivm committed:
changes

kunaltyagi commented:
We see a wall of code, but don't know why. Tests please

diivm committed:
[generate_bindings.py] stuff, not working as of yet

diivm committed:
[generate_bindings.py] fixed some stuff, needs cleaning

diivm committed:
[generate_bindings.py] cmake build without error!

diivm committed:
[generate_bindings.py] clean

diivm committed:
[CMakeLists] update path

diivm committed:
[generate.py] use utils

diivm committed:
[CMakeLists] path change

kunaltyagi commented on bindings/python/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@ project(bindings)
 find_package(pybind11)
 
-pybind11_add_module(pcl /home/divyanshu/Projects/active/pcl/bindings/python/pybind11/common/impl/point_types.cpp)
+pybind11_add_module(pcl /home/divyanshu/Projects/active/pcl/bindings/python/pybind11/common/include/pcl/impl/point_types.cpp)
```
There's `CMAKE_SOURCE_DIR`, `PROJECT_SOURCE_DIR`, `CMAKE_CURRENT_SOURCE_DIR`, etc. variables you can use. See [official docs](https://cmake.org/cmake/help/v3.5/manual/cmake-variables.7.html) for more details

diivm committed:
[generate.py] messy fix for anonymous gen and member definition

diivm committed:
[generate.py] add inclusion_directive

diivm committed:
[generate.py] fix cxx_method

diivm committed:
[generate.py] remove parameter_list from constructor

diivm committed:
[generate.py] some comments and minor fixes

diivm committed:
[CMakeLists] temp fix

- more robust change going on in another PR

diivm committed:
[test_generate.py] init file

diivm committed:
[generate.py] comment updation and minor refactoring

diivm committed:
[generate.py] list to tuple `in`s

divmadan commented:
@kunaltyagi @aponza
Here's this super, super, delayed PR for review.
Adding tests..

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        final.append(f"#include <{filename}>")
+        final.append("#include <pybind11/pybind11.h>")
+        # TODO: Maybe needed later
+        # final.append("#include<pybind11/stl.h>")
```
Maybe search for inclusion of stl? (Or just add it if it does work auto-magically)

diivm replied:
The pcl's header file will be included anyway, so any libraries included in the header will be included in the bound code too.

`final.append(f"#include <{filename}>")` L238

kunaltyagi replied:
I was referring to the TODO and how to resolve it:
* Either always add those headers
* Or detect if filename is in stdlib and then include it

kunaltyagi commented on bindings/python/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@ project(bindings)
 find_package(pybind11)
 
-pybind11_add_module(pcl common/py_point_types.cpp)
+pybind11_add_module(pcl ${CMAKE_CURRENT_SOURCE_DIR}/pybind11-gen/common/include/pcl/impl/point_types.cpp)
```
This file needs to be generated by another tool, right?

diivm replied:
`generate.py` will generate this file.
This CMakeLists is used for testing if the source compiles, no pybind11 related errors.

A more general version in #25:
https://github.com/divmadan/pcl/blob/86904e45a14611c296da483b5ce532a2e4505767/python/package/CMakeLists.txt#L7-L21

kunaltyagi reviewed (commented):
Will review more in depth soon.

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        self._skipped = []
+        self.inclusion_list = []
+        pybind_handle = self.skip  # handled by pybind11
+        other_function_handle = self.skip  # handled in another kind's function
```
Some new name ideas:
* handled_by_pybind
* handled_elsewhere
* no_need_to_handle

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
-        for line in linelist:
-            f.writelines(line)
-            f.writelines("\n")
+    # TODO: Not used, maybe remove
```
Seems simple enough to add later if needed. Go ahead an banish it :smile:

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
 
-in_struct = False
-module_linelist = []
+    def close(self):
```
Document all functions please

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+class bind:
+    def __init__(self, root):
+        self._state_stack = []
+        self.linelist = []
```
Segregate public and private variables (without and with a beginning underscore)

Also please add documentation for public members

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
-def read_json(filename):
-    with open(filename, "r") as f:
-        return json.load(f)
+class bind:
```
Documentation for the class, and how to use it

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+    def handle_node(self, item):
+        self.item = item
+        self.kind = self.item["kind"]
+        self.name = self.item["name"] if "name" in self.item else ""
```
```py
self.name = item.get("name", "")
```

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+            self.linelist.append(";")
 
+    def handle_node(self, item):
+        self.item = item
```
Might I recommend looking into namedtuple?

https://docs.python.org/3.5/library/collections.html#collections.namedtuple

diivm replied:
New issue created

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+            for sub_item in self.members:
+                self.handle_node(sub_item)
 
+        self.close()
```
Maybe choose a better name. `close` is pythonic for releasing ownership of a resource, like a dtor in C++

diivm replied:
Changed to `end_scope`

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
-    parameters_type = ""
-    parameters_type = ",".join(params for params in item["parameters_type"])
-    module_linelist.append(f".def(py::init<{parameters_type}>())")
+    def get_fields_from_anonymous(self, item):
```
This doesn't seem to use `self`. Please move out of class or mark as `@staticmethod`

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        fields = []
+        for sub_item in item["members"]:
+            if sub_item["kind"] in ("ANONYMOUS_UNION_DECL", "ANONYMOUS_STRUCT_DECL"):
+                for field in self.get_fields_from_anonymous(sub_item):
+                    fields.append(field)
```
```py
fields += get_fields_from_anonymous(sub_item)
```

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+            if sub_item["kind"] in ("ANONYMOUS_UNION_DECL", "ANONYMOUS_STRUCT_DECL"):
+                for field in self.get_fields_from_anonymous(sub_item):
+                    fields.append(field)
+            if sub_item["kind"] == "FIELD_DECL":
```
`elif`?

diivm replied:
More like a base condition. Updated:
https://github.com/divmadan/pcl/blob/81ed70c458d8cd35bb5610c9191009d559f6f65f/bindings/python/scripts/generate.py#L146-L152

kunaltyagi replied:
`elif`? since sub_item can't be of 2 diff kind at the same time

diivm replied:
Sure

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+    def get_fields_from_anonymous(self, item):
+        fields = []
+        for sub_item in item["members"]:
+            if sub_item["kind"] in ("ANONYMOUS_UNION_DECL", "ANONYMOUS_STRUCT_DECL"):
```
Are these the only 3 options available? (anon union/struct and field)

What about:
```cpp
struct rgb{
  union {
    struct { int8_t r, g, b, a};
    float rgba;
    struct details {  // named struct in anonymous struct
      ColorSpace space = ColorSpace::sRGB;
      // other things
    };
  };
};
```

What about public/private scope and member functions?

diivm replied:
Is this code C++11 compatible? https://godbolt.org/z/oKebnP

Also, [ an anonymous union may only have public non-static data members](https://stackoverflow.com/questions/17637392/anonymous-union-can-only-have-non-static-data-members-gcc-c)

kunaltyagi replied:
Didn't know that. This simplifies things. Do note it in the code (the restriction on anon union and structures)

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,2 @@
+from context import scripts
+import scripts.generate as generate
```
Tests incoming?

diivm replied:
Started

kunaltyagi reviewed (commented):
Partial review

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        return fields
 
+    def handle_struct_decl(self):
+        cxx_base_specifier_list = [
```
Maybe rename to `base_class_list` or something similar if this is what I think it is. Else... document what this is

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+            for sub_item in self.members
+            if sub_item["kind"] == "CXX_BASE_SPECIFIER"
+        ]
+        if cxx_base_specifier_list:
```
If you do things step by step, there should be no need of the if for string join operations

```py
# aim: add 'py::class_<struct_name{,base_class1...,base_classN}>(m, "struct_name")' to line

base_class_list_string = [str(class).replace("struct ", "") for class in base_class_list]
struct_details = ",".join([f"{struct_name}"] + base_class_list_string)
self.linelist.append(f"py::class_<{struct_details}>(m, "{struct_name}")
```

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        if cxx_base_specifier_list:
+            cxx_base_specifier_list = ",".join(cxx_base_specifier_list)
+            self.linelist.append(
+                f'py::class_<{self.name}, {cxx_base_specifier_list.replace("struct ", "").replace("pcl::", "")}>(m, "{self.name}")'
```
`pcl` or any namespace inside which we are?

Does pybind complain if `pcl::` is present? What about in case of `pcl::detail`?

diivm replied:
New issue created

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+                    )
 
+            # handle class methods
+            if sub_item["kind"] == "CXX_METHOD":
```
`elif`?

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        unnecessary_kind = self.skip  # unnecessary kind
+        unsure = self.skip  # unsure as to needed or not
+        self.kind_functions = {
+            "TRANSLATION_UNIT": unnecessary_kind,
```
Sort these (maybe group by the function and then sort each group)

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+        # TODO: Merge this and next block via a design updation
+        # handle anonymous structs, etc. as field declarations
+        for sub_item in self.members:
+            fields = self.get_fields_from_anonymous(sub_item)
+            for field in fields:
+                if field["element_type"] == "ConstantArray":
+                    # TODO: FIX: readwrite, not readonly
+                    self.linelist.append(
+                        f'.def_property_readonly("{field["name"]}", []({self.name}& obj) {{return obj.{field["name"]}; }})'  # float[ ' + f'obj.{sub_item["name"]}' + '.size()];} )'
+                    )
+                else:
+                    self.linelist.append(
+                        f'.def_readwrite("{field["name"]}", &{self.name}::{field["name"]})'
```
I get a feeling that this could be 2 separate functions:
* handle_anon
* handle_fields

and you'd just call them like:
```python
for elem in self.members:
    self.kind_functions[elem["kind"](elem)
```

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
-    module_linelist.append("}")
+            # handle field declarations
+            if sub_item["kind"] == "FIELD_DECL":
+                if sub_item["element_type"] == "ConstantArray":
```
Could you give an example of constant array?

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+            self.linelist.append(f'py::class_<{self.name}>(m, "{self.name}")')
 
+    def handle_inclusion_directive(self):
+        # TODO: update blacklist
```
Why a blacklist?

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,284 @@
+    def handle_inclusion_directive(self):
+        # TODO: update blacklist
+        # blacklist = ["pcl/memory.h", "pcl/pcl_macros.h",]
+        if self.name.startswith("pcl"):
```
Is it possible to factor out `"pcl"` to the class ctor?

kunaltyagi reviewed (commented):
4 functions left

diivm committed:
[generate.py] new names

diivm committed:
[generate.py] remove `get_prev_depth_node`

diivm committed:
[generate.py]  add docstrings - 1

diivm committed:
[generate.py] add libraries everytime

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,318 @@
-    "macro": handle_macro,
-}
+    def handle_final(self, filename, module_name):
+        final = []
```
No need to append when you can directly put them in the array

diivm replied:
Taken care of.

diivm committed:
[generate.py] unnecessary if

diivm committed:
[generate.py] close() -> end_scope() rename

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,318 @@
+        self.kind_functions = {
+            "TRANSLATION_UNIT": no_need_to_handle,
+            "NAMESPACE": self.handle_namespace,
+            "NAMESPACE_REF": handled_elsewhere,  # in (handle_constructor)
+            "STRUCT_DECL": self.handle_struct_decl,
+            "CXX_BASE_SPECIFIER": handled_elsewhere,  # in (handle_struct_decl)
+            "CXX_METHOD": handled_elsewhere,  # in (handle_struct_decl)
+            "VAR_DECL": handled_by_pybind,
+            "TYPE_REF": handled_elsewhere,  # in (handle_constructor)
+            "CONSTRUCTOR": self.handle_constructor,
+            "PARM_DECL": handled_elsewhere,  # in (handle_constructor)
+            "CALL_EXPR": handled_by_pybind,
+            "UNEXPOSED_EXPR": unsure,
+            "MEMBER_REF_EXPR": unsure,
+            "DECL_REF_EXPR": unsure,
+            "FIELD_DECL": handled_elsewhere,  # in (handle_struct_decl)
+            "MEMBER_REF": handled_by_pybind,
+            "CLASS_TEMPLATE": self.skip,  # self.handle_class_template
+            "TEMPLATE_NON_TYPE_PARAMETER": handled_elsewhere,  # in (handle_class_template)
+            "FUNCTION_TEMPLATE": self.skip,  # to be added later
+            "ANONYMOUS_UNION_DECL": handled_elsewhere,  # in (handle_struct_decl) via get_fields_from_anonymous
+            "ALIGNED_ATTR": no_need_to_handle,
+            "INTEGER_LITERAL": unsure,
+            "ANONYMOUS_STRUCT_DECL": handled_elsewhere,  # in (handle_struct_decl) via get_fields_from_anonymous
+            "COMPOUND_STMT": no_need_to_handle,
+            "FLOATING_LITERAL": unsure,
+            "BINARY_OPERATOR": no_need_to_handle,
+            "ARRAY_SUBSCRIPT_EXPR": handled_by_pybind,
+            "CXX_THROW_EXPR": handled_by_pybind,
+            "FRIEND_DECL": unsure,
+            "FUNCTION_DECL": unsure,
+            "INIT_LIST_EXPR": no_need_to_handle,
+            "RETURN_STMT": handled_by_pybind,
+            "OVERLOADED_DECL_REF": unsure,
+            "UNARY_OPERATOR": no_need_to_handle,
+            "IF_STMT": no_need_to_handle,
+            "OBJ_BOOL_LITERAL_EXPR": unsure,
+            "INCLUSION_DIRECTIVE": self.handle_inclusion_directive,
+            "MACRO_DEFINITION": unsure,
+            "MACRO_INSTANTIATION": unsure,
```
group and sort them based on some intuitive logic

diivm committed:
[generate.py] make `get_fields_from_anonymous` static

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -144,7 +145,7 @@ def get_fields_from_anonymous(self, item):
         for sub_item in item["members"]:
             if sub_item["kind"] in ("ANONYMOUS_UNION_DECL", "ANONYMOUS_STRUCT_DECL"):
-                for field in self.get_fields_from_anonymous(sub_item):
+                for field in bind.get_fields_from_anonymous(sub_item):
```
`bind` is a reserved word in Python?

diivm replied:
I don't think so
```py
>>> import keyword
>>> for k in keyword.kwlist:
...     print(k)
... 
False
None
True
and
as
assert
break
class
continue
def
del
elif
else
except
finally
for
from
global
if
import
in
is
lambda
nonlocal
not
or
pass
raise
return
try
while
with
yield


```

kunaltyagi replied:
Oh cool. PS: why change from `self` to `bind`?

diivm replied:
Changed to `@staticmethod` as suggested by you, so no `self`.

kunaltyagi replied:
```python
class A:
    def __init__(self):
        self.say_hi()
    @staticmethod
    def say_hi():
        print("I'm a static method, short and stout")

A()
A.say_hi()
```

diivm committed:
[generate.py] `get_fields_from_anonymous` better internals

diivm committed:
[generate.py] extract generation for testing calls

diivm committed:
[test_generate.py] intial setup and some tests

diivm committed:
[generate.py] elif fix

diivm committed:
[generate.py] separate pybind line list

diivm committed:
[generate.py] fix hardcoded filepath

diivm committed:
[test_generate.py] fix intial_lines problem, new function

diivm committed:
[generate.py] add template support

bugs encountered:
- extra constructor at the beginning of file pcl_base.cpp gives rise to standalone .def
- inclusion of cpp files: header files must be included, not the cpp file
- namespace pcl should be in all code
- `;` at the end of class not generated if no defs

kunaltyagi commented:
What changed in the force push? The PR base or the something in code was added since last review?

diivm committed:
[generate.py] minor fix, use class_name

kunaltyagi commented:
This PR is super confusing. Should this be closed or something else?

divmadan commented:
> This PR is super confusing. Should this be closed or something else?

@kunaltyagi 
Yeah, just taking in your reviews and creating commits/issues out of them. Will close in a bit.

divmadan commented:
Remaining reviews transferred to #41 
Closing this in favour of #41
